### PR TITLE
Attempt to fix issue 43

### PR DESCRIPTION
**Pull request by Skgland**
maps error for all remove_dir_all calls where the result is not ignored (all but one) to the added Custom error type to also display the directory path

EDIT: also do this for file removals

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -50,6 +50,58 @@ fn strip_verbatim_from_prefix(prefix: &PrefixComponent<'_>) -> Option<PathBuf> {
...
+    let tested = format!(
+        "{}",
+        improve_remove_error(
+            std::io::Error::from(std::io::ErrorKind::PermissionDenied),
+            path
+        )
+    );
+    assert_eq!(expected, tested);
+
+    let expected = "Custom { kind: PermissionDenied, error: RemoveError { kind: PermissionDenied, path: \"test/path\" } }";
```
Comment (emilyalbini) [inline src/utils.rs]: The debug representation is not guaranteed to be stable, and I wouldn't test it.

File: src/utils.rs [hunk 2]
```diff
@@ -50,6 +50,58 @@ fn strip_verbatim_from_prefix(prefix: &PrefixComponent<'_>) -> Option<PathBuf> {
...
+        error.kind(),
+        RemoveError {
+            kind: error.kind(),
+            path: path.display().to_string(),
+        },
+    )
+}
+
+#[test]
+fn custom_remove_error() {
```
Comment (emilyalbini) [inline src/utils.rs]: Can you move the test to a `mod tests { ... }`?

File: src/utils.rs [hunk 3]
```diff
@@ -50,6 +50,58 @@ fn strip_verbatim_from_prefix(prefix: &PrefixComponent<'_>) -> Option<PathBuf> {
     Some(ret)
 }
 
+#[derive(Debug)]
+struct RemoveError {
+    kind: std::io::ErrorKind,
+    path: String,
```
Comment (emilyalbini) [inline src/utils.rs]: Let's store it as a `PathBuf`.

File: src/utils.rs [hunk 4]
```diff
@@ -50,6 +50,58 @@ fn strip_verbatim_from_prefix(prefix: &PrefixComponent<'_>) -> Option<PathBuf> {
...
+impl std::fmt::Display for RemoveError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!(
+            "failed to remove '{}' : {:?}",
+            self.path, self.kind
+        ))
+    }
+}
+
+pub(crate) fn improve_remove_error(error: std::io::Error, path: &Path) -> std::io::Error {
```
Comment (emilyalbini) [inline src/utils.rs]: I feel like it'd be better to create custom `remove_file` and `remove_dir_all` functions than using `.map_err` all over the codebase.

**Review by emilyalbini — COMMENTED** [review]
Thanks for working on this! Left some feedback.

**Comment by shepmaster** [comment]
This may be reimplementing https://crates.io/crates/fs-err

**Comment by Skgland** [comment]
Did't know about `fs-err` until now. Though that wouldn't work for `remove_dir_all` as this project is not using `std::fs::remove_dir_all`, probably because of <https://github.com/rust-lang/rust/issues/29497>.
So that wrapper of the two would still be necessary.

**Comment by Skgland** [comment]
Though it might be reasonable to also get better errors for other Filesystem operations using `fs-err` only leaving `remove_dir_all` like this.

**Comment by Skgland** [comment]
Opened PR #47 to expand on this using `fs-err` for all Filesystem operations where possible.

**Comment by emilyalbini** [comment]
I feel like this approach is better than #47: bringing another dependency just for this is not really needed.

Thanks for working on the PRs! Merging.

**Pull request merged**